src/OrleansTestingHost/AppDomainSiloHost.cs

```diff
@@ -87,4 +101,87 @@ public void Shutdown()
             silo.Shutdown();
         }
     }
+
+    /// <summary>
+    /// Test hook functions for white box testing.
+    /// NOTE: this class has to and will be removed entirely. This requires the tests that currently rely on it, to assert using different mechanisms, such as with grains.
+    /// </summary>
+    internal class AppDomainTestHooks : MarshalByRefObject
+    {
+        private readonly Silo silo;
+
+        public AppDomainTestHooks(Silo silo)
+        {
+            this.silo = silo;
+        }
+
+        internal IBootstrapProvider GetBootstrapProvider(string name)
+        {
+            IBootstrapProvider provider = silo.BootstrapProviders.First(p => p.Name.Equals(name));
+            return CheckReturnBoundaryReference("bootstrap provider", provider);
+        }
+
+        /// <summary>Find the named storage provider loaded in this silo. </summary>
+        internal IStorageProvider GetStorageProvider(string name) => CheckReturnBoundaryReference("storage provider", (IStorageProvider)silo.StorageProviderManager.GetProvider(name));
+
+        private static T CheckReturnBoundaryReference<T>(string what, T obj) where T : class
+        {
+            if (obj == null) return null;
+            if (obj is MarshalByRefObject || obj is ISerializable)
+            {
+                // Referernce to the provider can safely be passed across app-domain boundary in unit test process
+                return obj;
+            }
+            throw new InvalidOperationException(
+                $"Cannot return reference to {what} {TypeUtils.GetFullName(obj.GetType())} if it is not MarshalByRefObject or Serializable");
+        }
+
+        public IDictionary<GrainId, IGrainInfo> GetDirectoryForTypeNamesContaining(string expr)
+        {
+            var x = new Dictionary<GrainId, IGrainInfo>();
+            foreach (var kvp in ((LocalGrainDirectory)silo.LocalGrainDirectory).DirectoryPartition.GetItems())
+            {
+                if (kvp.Key.IsSystemTarget || kvp.Key.IsClient || !kvp.Key.IsGrain)
+                    continue;// Skip system grains, system targets and clients
+                if (((Catalog)silo.Catalog).GetGrainTypeName(kvp.Key).Contains(expr))
+                    x.Add(kvp.Key, kvp.Value);
+            }
+            return x;
+        }
+        
+        // store silos for which we simulate faulty communication
+        // number indicates how many percent of requests are lost
+        private ConcurrentDictionary<IPEndPoint, double> simulatedMessageLoss;
+        private readonly SafeRandom random = new SafeRandom();
+
+        internal void BlockSiloCommunication(IPEndPoint destination, double lossPercentage)
+        {
+            if (simulatedMessageLoss == null)
+                simulatedMessageLoss = new ConcurrentDictionary<IPEndPoint, double>();
+
+            simulatedMessageLoss[destination] = lossPercentage;
+
+            var mc = (MessageCenter)silo.LocalMessageCenter;
```

Comment:
similarly, because the code is now moved to a non-nested class, it can no longer access private fields, so the casting would have to either be done in the constructor or here (or expose implementation details that we probably don't want to do for code that should eventually be removed).